src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpUseNotPatternDiagnosticAnalyzer.cs

```diff
@@ -64,28 +64,46 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
             if (!styleOption.Value)
                 return;
 
-            // Look for the form: !(x is Y y)
-            if (!(node is PrefixUnaryExpressionSyntax
+            // Look for the form: !(x is Y y) and !(x is const)
+            if (node is PrefixUnaryExpressionSyntax
                 {
                     Operand: ParenthesizedExpressionSyntax
                     {
                         Expression: IsPatternExpressionSyntax
                         {
-                            Pattern: DeclarationPatternSyntax,
+                            Pattern: DeclarationPatternSyntax or ConstantPatternSyntax
                         } isPattern,
                     },
-                } notExpression))
+                } notExpression)
             {
+                syntaxContext.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    isPattern.IsKeyword.GetLocation(),
+                    styleOption.Notification.Severity,
+                    ImmutableArray.Create(notExpression.GetLocation()),
+                    properties: null));
                 return;
             }
 
-            // Put a diagnostic with the appropriate severity on `is` keyword.
-            syntaxContext.ReportDiagnostic(DiagnosticHelper.Create(
-                Descriptor,
-                isPattern.IsKeyword.GetLocation(),
-                styleOption.Notification.Severity,
-                ImmutableArray.Create(notExpression.GetLocation()),
-                properties: null));
+            // Look for the form: !(x is Y)
+            if (node is PrefixUnaryExpressionSyntax
+                {
+                    Operand: ParenthesizedExpressionSyntax
+                    {
+                        Expression: BinaryExpressionSyntax
+                        {
+                            Right: TypeSyntax
+                        } isExpression,
+                    },
+                } notIsExpression && isExpression.IsKind(SyntaxKind.IsExpression))
```

Comment:
```suggestion
                    Operand: ParenthesizedExpressionSyntax { Expression: BinaryExpressionSyntax(SyntaxKind.IsExpression) },
                })
```

Reply:
I don't understand how `Expression: BinaryExpressionSyntax(SyntaxKind.IsExpression)` is supposed to work.

The only way to check if it's an is-expression using pattern matching I came up with is
```c#
Expression: BinaryExpressionSyntax { RawKind: (int)SyntaxKind.IsExpression }
```
but I'm not sure how it looks.

Reply:
We should have a deconstruct member here to allow a positional pattern to match type and kind.  I'm not in front of the computer, but I can try to find later. 

Or maybe @alrz or @Youssef1313 know where it is?

Reply:
https://github.com/dotnet/roslyn/blob/5c707a8c0bb35aabceca2f273b14fedea8bd1c5f/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxNodeExtensions.cs#L24-L27